### PR TITLE
Expose Merkle Proof

### PR DESCRIPTION
**Pull request by samoht**
Forward port #1583

The port is a bit more involved with structured keys but I think the result is quite nice. There are a few quirks in the current API organisation that I plan to tackle in a separate PR.

### CODE REVIEW

**Comment by samoht** [comment]
(I'll fix the CHANGELOG in another PR which will fix all the 2.9 items in one go)

File: src/irmin/tree.ml [hunk 1]
```diff
@@ -892,7 +972,7 @@ module Make (P : Backend.S) = struct
       | None -> t.info.findv_cache <- Some (StepMap.singleton step v)
       | Some m -> t.info.findv_cache <- Some (StepMap.add step v m)
 
-    let findv ~cache ctx t step =
+    let findv_aux ~cache ~value_of_key ~return ~bind ctx t step =
```
Comment (maiste) [inline src/irmin/tree.ml:975]: What is the reason to introduce a `return` and a `bind` parameter in this function? I'm not sure to understand their purpose...

File: src/irmin/tree.ml [hunk 2]
```diff
@@ -1731,7 +1875,7 @@ module Make (P : Backend.S) = struct
                            | _, Remove -> None)
                   | Map m -> StepMap.to_seq m
                   | Value (_, _, None) -> Seq.empty
-                  | Key _ | Pruned _ ->
+                  | Key _ | Pruned _ | Portable _ ->
                       (* [n.v = (Key _ | Pruned _)] is excluded above. *)
```
Comment (maiste) [inline src/irmin/tree.ml:1879]:
```suggestion
                      (* [n.v = (Key _ | Pruned _ | Portable _)] is excluded above. *)
```

File: src/irmin/tree.ml [hunk 3]
```diff
@@ -1742,7 +1886,7 @@ module Make (P : Backend.S) = struct
               | Map x, _ -> add_node_map n x k
               | Value (_, v, None), None | _, Some v -> add_node n v k
               | Value (_, v, Some um), _ -> add_updated_node n v um k
-              | (Key _ | Pruned _), _ ->
+              | (Key _ | Pruned _ | Portable _), _ ->
                   (* [n.v = (Key _ | Pruned _)] is excluded above. *)
```
Comment (maiste) [inline src/irmin/tree.ml:1890]:
```suggestion
                  (* [n.v = (Key _ | Pruned _ | Portable _)] is excluded above. *)
```

File: src/irmin/tree.ml (hunk 1 above)
Comment (samoht) [reply src/irmin/tree.ml:975]: Because it's used in `Proof.findv` which isn't in the Lwt monad.

Comment (maiste) [reply src/irmin/tree.ml:975]: Oh, I see! Thanks for the explanation :+1:

**Comment by samoht** [comment]
I'll split this PR a bit to simplify the review.

**Comment by samoht** [comment]
Now rebased on top of #1612

**Review by Ngoguey42 — COMMENTED** [review]
I'm still in the process of reviewing it, here is a temporary review :)

**Comment by icristescu** [comment]
The forward port for merkle proofs is done, so closing this.

**Pull request closed without merging**